### PR TITLE
feat: WebSocket upgrade support

### PR DESCRIPTION
**Pull request by nyannyacha**
## What kind of change does this PR introduce?

Feature

## Description

This PR is to make it possible for the (user/main) worker to upgrade incoming requests to the WebSocket protocol.

### Breaking changes

* `EdgeRuntime.applyConnectionWatcher` is replaced by `EdgeRuntime.applySupabaseTag`
  This change is a result of the need to track multiple pieces of data.

### CODE REVIEW

**Comment by nyannyacha** [comment]
Note: Tests are passed, but there is no test for the upgrade in the WebSocket. I need to write a test.

**Comment by nyannyacha** [comment]
ready for review 😋
cc @andreespirela @laktek

File: crates/sb_core/js/01_http.js [hunk 1]
```diff
@@ -0,0 +1,554 @@
...
+      false,
+    );
+
+    const respondWith = createRespondWith(
+      this,
+      streamRid,
+      abortController,
+    );
+
+    return { request, respondWith, streamRid };
```
Comment (nyannyacha) [inline crates/sb_core/js/01_http.js:174]: Note: This is the only purpose this file exists for.
We need `streamRid` for protocol upgrades, but Deno hasn't exposed it at all.

File: crates/base/Cargo.toml [hunk 2]
```diff
@@ -6,6 +6,7 @@ edition = "2021"
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
+base_share = { version = "0.1.0", path = "../base_share" }
```
Comment (laktek) [inline crates/base/Cargo.toml:9]: Wonder if we should call it `base_share` or can we make it `http_exts` or `http_utils`? Want to keep the crates focused and less confusing. Having a `base_share` crate may cause confusions for other contributors.

Comment (nyannyacha) [reply crates/base/Cargo.toml:9]: @laktek 
Yeah, `http_utils` would be better.
I’ll reflect that 😁

Comment (nyannyacha) [reply crates/base/Cargo.toml]: done

File: crates/base/src/rt_worker/worker_ctx.rs [hunk 3]
```diff
@@ -83,43 +89,98 @@ async fn handle_request(
...
+
+                    if let Some(mut watcher) = conn_watch {
+                        if watcher.wait_for(|it| *it == ConnSync::Recv).await.is_err() {
+                            error!("cannot track outbound connection correctly");
+                        }
                     }
-                }
 
-                drop(parts);
+                    drop(parts);
```
Comment (andreespirela) [inline crates/base/src/rt_worker/worker_ctx.rs:139]: Is this necessary?. Does it drop it at the end of the statement anyways?

Comment (nyannyacha) [reply crates/base/src/rt_worker/worker_ctx.rs:139]: I think I just paranoidly put it in there 🙄

Comment (nyannyacha) [reply crates/base/src/rt_worker/worker_ctx.rs]: done

**Review by laktek — APPROVED** [review]

**Pull request merged**